File: 2022/day1.jl
using Test

function part_one(input)
    elves = split.(split(input, "\n\n"), "\n")
    per_elf = map(e -> sum(parse.(Int64, e)), elves)
    return maximum(per_elf)
end

function part_two(input)
    elves = split.(split(input, "\n\n"), "\n")
    per_elf = map(e -> sum(parse.(Int64, e)), elves)
    return sum(sort(per_elf)[end-2:end])
end

function test()
    example = """1000
2000
3000

4000

5000
6000

7000
8000
9000

10000"""

    @test part_one(example) == 24000
    @test part_two(example) == 45000
end

File: ruby/2018/day1.rb
# frozen_string_literal: true

require 'set'
require './commons/assert'

def part_one(input)
  input.split("\n").map(&:to_i).sum
end

def part_two(input)
  freqs = input.split("\n").map(&:to_i)
  current = 0
  found = Set[0]

  loop do
    freqs.each do |f|
      current += f
      return current if found.include?(current)

      found.add(current)
    end
  end
end

def test
  example = %(+1
-2
+3
+1)

  assert part_one(example), 3
  assert part_two(example), 2
  0
end
